pkg/kudoctl/cmd/uninstall.go

```diff
@@ -0,0 +1,130 @@
+package cmd
+
+import (
+	"fmt"
+
+	"github.com/kudobuilder/kudo/pkg/kudoctl/clog"
+	"github.com/kudobuilder/kudo/pkg/kudoctl/env"
+	"github.com/kudobuilder/kudo/pkg/kudoctl/util/kudo"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	uninstallExample = `kubectl kudo uninstall --instance flink`
+)
+
+type uninstallOptions struct {
+	InstanceName string
+	Purge        bool
+}
+
+type uninstallCmd struct{}
+
+func (cmd *uninstallCmd) run(options uninstallOptions, settings *env.Settings) error {
+	kc, err := kudo.NewClient(settings.Namespace, settings.KubeConfig)
+	clog.V(3).Printf("acquiring kudo client")
+	if err != nil {
+		clog.V(3).Printf("failed to acquire kudo client")
+		return fmt.Errorf("failed to acquire kudo client: %w", err)
+	}
+
+	return cmd.uninstall(kc, options.InstanceName, options.Purge, settings)
+}
+
+func (cmd *uninstallCmd) uninstall(kc *kudo.Client, instanceName string, purge bool, settings *env.Settings) error {
+	instance, err := kc.GetInstance(instanceName, settings.Namespace)
+	if err != nil {
+		return fmt.Errorf("failed to verify if instance already exists: %w", err)
+	}
+
+	if instance == nil {
+		return fmt.Errorf("instance %s in namespace %s does not exist in the cluster", instanceName, settings.Namespace)
+	}
+
+	operatorVersionName := instance.Spec.OperatorVersion.Name
+	operatorVersionNamespace := settings.Namespace
+	if instance.Spec.OperatorVersion.Namespace != "" {
+		operatorVersionNamespace = instance.Spec.OperatorVersion.Namespace
+	}
+
+	err = kc.DeleteInstance(instanceName, settings.Namespace)
+	if err != nil {
+		return err
+	}
+
+	if purge {
```

Comment:
I am not sure if we should not have some safety net here. For now, I would allow deleting OV and O only, if there is no instance using them otherwise it's gonna be ugly.

We can optionally in other PR add another parameter `--force` that will just delete OV, O and all related instances if people find that useful

Reply:
That's a very good point. I'm not sure though how to figure out all instances that belong to an `OperatorVersion`. We could iterate over all instances and see if more than one reference the same `OperatorVersion`? This should be done _before_ deleting the particular instance.

Reply:
> For now, I would allow deleting OV and O only, if there is no instance using them otherwise it's gonna be ugly.

Agreed. IMO, force-deleting an OV or O would even justify a new subcommand for it (e.g. `kubectl kudo unsafe delete $ov`).

In this sentiment, does it make sense to always do what @alenkacz suggested, and drop the `purge` flag? i.e., always delete the OV if there are no instances using it and always delete the O if there are no OVs using it?

Reply:
I think there is no good (meaning smart) way to do that. You have to pull all instances and compare the `Spec.OperatorVersion` exactly as you said. We had OV name in labels before but we had to remove it.I think for CLI though this should be fine (up to very very big clusters :))

Reply:
I see. This is probably outside the context of this PR, but would it be feasible for the controller-manager to maintain indexes for "instances depending on OV" and "OVs depending on O"? That way it could have a smarter way to check dependencies that is not iterating resource by resource?

OV index:
```json
{"cassandra-0.1.0": ["c-instance-0", "c-instance-1"],
 "cassandra-0.2.0": ["c-instance-2", "c-instance-3"],
 "kafka-0.1.0": ["k-instance-0"]}
```
O index:
```json
{"cassandra": ["cassandra-0.1.0", "cassandra-0.2.0", "cassandra-0.3.0"],
 "kafka": ["kafka-0.1.0"]}
```

Reply:
@alenkacz Yes, that's my current approach, going through all instances and checking their `Instance.Spec.OperatorVersion.Name`.
@mpereira I'd like to avoid non-trivial or "magic" behavior and ideally provide idempotent operations -- even though `uninstall` currently is not. This is something that wouldn't work when removing OV/O if there's only a single instance left. But this raises some interesting questions. At the moment the CLI can only list instances but not OV/O. But one can already only install OV/O without installing an instance. We should add commands to list OV/O using the CLI. With that we could add flags to target OV/O instead of instances for certain operations, including `uninstall`. E.g.
```
kudo get operatorversions
kudo uninstall --operatorversion %operatorversion%
```

Reply:
Thinking about it a bit more, I think we shouldn't remove `OperatorVersion` and `Operator` for now and add this functionality later when we are also able to list OV/O from the CLI. As long as the KUDO CLI doesn't provide a way to inform a user about this, we should expect the user to use `kubectl delete` to remove these resources.

Reply:
I removed the `purge` flag. Only instances are deleted. We should discuss the handling of `OperatorVersion` and `Operator` and the implications of this at a later point.